File: apps/mobile/ios/Sources/Flagship/BootAuthorization.swift
import Foundation
import CryptoKit

/// Mirrors @flagship/protocol's auth canonicalization. Keep in sync with
/// `packages/protocol/src/auth.ts`.
public struct BootChallenge: Codable {
    public let serverId: String
    public let nonce: Data
    public let issuedAt: Int64
}

public struct BootAuthorization {
    public static func sign(challenge: BootChallenge, with bak: Curve25519.Signing.PrivateKey) throws -> Data {
        let canonical = canonicalBytes(challenge)
        return try bak.signature(for: canonical)
    }

    private static func canonicalBytes(_ c: BootChallenge) -> Data {
        let s = "flagship/boot/v1|\(c.serverId)|\(c.nonce.hexString)|\(c.issuedAt)"
        return Data(s.utf8)
    }
}

extension Data {
    var hexString: String {
        map { String(format: "%02x", $0) }.joined()
    }
}

File: apps/mobile/android/app/src/main/java/com/flagship/keystore/BootAuthorization.kt
package com.flagship.keystore

import com.google.crypto.tink.signature.Ed25519PrivateKeyManager
import com.google.crypto.tink.subtle.Ed25519Sign

/**
 * Mirrors @flagship/protocol's auth canonicalization. Keep in sync with
 * `packages/protocol/src/auth.ts`.
 */
data class BootChallenge(
    val serverId: String,
    val nonce: ByteArray,
    val issuedAt: Long
)

object BootAuthorization {
    fun sign(challenge: BootChallenge, bakSeed: ByteArray): ByteArray {
        val signer = Ed25519Sign(bakSeed)
        return signer.sign(canonicalBytes(challenge))
    }

    private fun canonicalBytes(c: BootChallenge): ByteArray {
        val s = "flagship/boot/v1|${c.serverId}|${c.nonce.toHex()}|${c.issuedAt}"
        return s.toByteArray(Charsets.UTF_8)
    }

    private fun ByteArray.toHex(): String =
        joinToString("") { "%02x".format(it) }
}
